shared/ssl.sh

```diff
@@ -130,38 +195,47 @@ fi
 [ -s "${CSRFILE}" ] && error "certificate request '${CSRFILE}' already exists"
 openssl genrsa -out "${PRIVKEYFILE}" 4096
 openssl rsa -in "${PRIVKEYFILE}" -pubout -out "${PUBKEYFILE}"
+# shellcheck disable=SC2086 # $CERTEXTENSIONS shouldn't be quoted
 openssl req -new -key "${PRIVKEYFILE}" -out "${CSRFILE}" -subj "${CERTSUBJECT}" ${CERTEXTENSIONS}
+[ -f "${ALTNAMEFILE}" ] && rm "${ALTNAMEFILE}"
 
 ### Submit CSR and fail gracefully on certain error conditions
-output="$(curl ${CURLFLAGS} -X PUT -H "Content-Type: text/plain" \
-               --data-binary "@${CSRFILE}" "${CA}/certificate_request/${CERTNAME}")"
-if [ -n "${output}" ]; then
+CSRREQ=$(cat <<EOF
+PUT ${CA}/certificate_request/${CERTNAME} HTTP/1.0
+Content-Length: $(wc -c < "${CSRFILE}")
+Content-Type: text/plain
+
+$(cat "${CSRFILE}")
+EOF
+)
+
+if ! output=$(httpsreq "$CSRREQ"); then
     cert_already_exists="${CERTNAME} already has a requested certificate; ignoring certificate request"
     altnames_disallowed="CSR '${CERTNAME}' contains subject alternative names*which are disallowed*"
     case "${output}" in
-        $cert_already_exists) error "unsigned CSR for '${CERTNAME}' already exists on CA" ;;
-        $altnames_disallowed) error "DNS Alt Names not allowed by the CA" ;;
+        "$cert_already_exists") error "unsigned CSR for '${CERTNAME}' already exists on CA" ;;
```

Comment:
The addition of quotes here on the `case` cases might not be a refactor. I remember it being very fiddly/specific.

Reply:
I can double check, but the shellcheck warning says that it's treated as a glob if it's not quoted.

Reply:
It has worked successfully for us already, and I remember quotes changing the behavior while I developed it.

If it works with quotes then let's keep them though!

Reply:
Oh and it needs to be a glob since we're pattern matching here.

Right?

Eh double check 😄

Reply:
Yeah I think the quotes causes the globs in the variables to not expand, which we're relying on.

Reply:
https://github.com/koalaman/shellcheck/wiki/SC2254

Reply:
That example doesn't use globbing

Reply:
We want globbing because the error message includes text that we can't know/hardcode to match, like other alt names.

Reply:
Yeah everything you're saying is correct if we want LITERAL matching but we want GLOBBING because we're matching on error messages that include unknown text.  The variable `altnames_disallowed` contains `*` globs, which won't be expanded properly by the `case` if there are quotes around the variable.

All the shellcheck stuff says "if you don't want globbing" then add quotes, but we do want globbing.

Reply:
Ah yeah, completely missed the embedded `*` in the `altnames_disallowed` variable - thanks. 

I don't see any globs in `cert_already_exists` though.